src/guacenc/display-flatten.c

```diff
@@ -0,0 +1,160 @@
+    int a_depth = guacenc_display_get_depth(__qsort_display, layer_a);
+    int b_depth = guacenc_display_get_depth(__qsort_display, layer_b);
+    if (b_depth != a_depth)
+        return b_depth - a_depth;
+
+    /* Order such that sibling layers are adjacent */
+    if (layer_b->parent_index != layer_a->parent_index)
+        return layer_b->parent_index - layer_a->parent_index;
+
+    /* Order sibling layers according to descending Z */
+    return layer_b->z - layer_a->z;
+
+}
+
+int guacenc_display_flatten(guacenc_display* display) {
+
+    int i;
+    guacenc_layer* render_order[GUACENC_DISPLAY_MAX_LAYERS];
+
+    /* Copy list of layers within display */
+    memcpy(render_order, display->layers, sizeof(render_order));
+
+    /* Sort layers by depth, parent, and Z */
+    __qsort_display = display;
+    qsort(render_order, GUACENC_DISPLAY_MAX_LAYERS, sizeof(guacenc_layer*),
+            guacenc_display_layer_comparator);
+
+    /* Reset layer frame buffers */
+    for (i = 0; i < GUACENC_DISPLAY_MAX_LAYERS; i++) {
+
+        /* Pull current layer, ignoring unallocated layers */
+        guacenc_layer* layer = render_order[i];
+        if (layer == NULL)
+            continue;
+
+        /* Get source buffer and destination frame buffer */
+        guacenc_buffer* buffer = layer->buffer;
+        guacenc_buffer* frame = layer->frame;
+
+        /* Reset frame contents */
+        guacenc_buffer_copy(frame, buffer);
+
+    }
+
+    /* Render each layer, in order */
+    for (i = 0; i < GUACENC_DISPLAY_MAX_LAYERS; i++) {
+
+        /* Pull current layer, ignoring unallocated layers */
+        guacenc_layer* layer = render_order[i];
+        if (layer == NULL)
+            continue;
+
+        /* Skip fully-transparent layers */
+        if (layer->opacity == 0)
+            continue;
+
+        /* Ignore layers without a parent */
+        int parent_index = layer->parent_index;
+        if (parent_index == GUACENC_LAYER_NO_PARENT)
+            continue;
+
+        /* Retrieve parent layer, ignoring layers with invalid parents */
+        guacenc_layer* parent = guacenc_display_get_layer(display, parent_index);
+        if (parent == NULL)
+            continue;
+
+        /* Get source and destination frame buffer */
+        guacenc_buffer* src = layer->frame;
+        guacenc_buffer* dst = parent->frame;
+
+        /* Ignore layers with empty buffers */
+        cairo_surface_t* surface = src->surface;
+        if (surface == NULL)
+            continue;
+
+        /* Ignore if parent has no pixels */
+        cairo_t* cairo = dst->cairo;
+        if (cairo == NULL)
+            continue;
+
+        /* Render buffer to layer */
+        cairo_reset_clip(cairo);
+        cairo_rectangle(cairo, layer->x, layer->y, src->width, src->height);
+        cairo_clip(cairo);
+
+        cairo_set_source_surface(cairo, surface, layer->x, layer->y);
+        cairo_paint_with_alpha(cairo, layer->opacity / 255.0);
```

Comment:
I'm not sure I agree. All color components in Guacamole, including alpha, are documented as having a maximum of `0xFF` (255) and a minimum of `0x00`, and that maximum is a common convention within graphics. In Cairo, these range between `0.0` and `1.0`. We're simply converting the ranges here.